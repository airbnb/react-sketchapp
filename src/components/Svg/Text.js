import React from 'react';
import PropTypes from 'prop-types';
import { textProps } from './props';

module.exports = class Text extends React.Component {
  static propTypes = textProps;

  static childContextTypes = {
    isInAParentText: PropTypes.bool,
  };

  getChildContext() {
    return {
      isInAParentText: true,
    };
  }

  getContextTypes() {
    return {
      isInAParentText: PropTypes.bool,
    };
  }

  render() {
    const { children, ...rest } = this.props;
    return <svg_text {...rest}>{children}</svg_text>;
  }
};
