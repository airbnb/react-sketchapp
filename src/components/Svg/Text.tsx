import * as React from 'react';
import * as PropTypes from 'prop-types';
import { textProps } from './props';

type Props = PropTypes.InferProps<typeof textProps>;

export class Text extends React.Component<Props> {
  static propTypes = textProps;

  static childContextTypes = {
    isInAParentText: PropTypes.bool,
  };

  getChildContext() {
    return {
      isInAParentText: true,
    };
  }

  getContextTypes() {
    return {
      isInAParentText: PropTypes.bool,
    };
  }

  render() {
    const { children, ...rest } = this.props;
    return <svg_text {...rest}>{children}</svg_text>;
  }
}
