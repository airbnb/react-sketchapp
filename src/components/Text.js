/* @flow */
import React from 'react';
import PropTypes from 'prop-types';
import StyleSheet from '../stylesheet';
import TextStylePropTypes from './TextStylePropTypes';
import ViewStylePropTypes from './ViewStylePropTypes';
import ResizingConstraintPropTypes from './ResizingConstraintPropTypes';

const propTypes = {
  // TODO(lmr): do some nice warning stuff like RN does
  style: PropTypes.oneOfType([
    PropTypes.shape({ ...ViewStylePropTypes, ...TextStylePropTypes }),
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.shape({ ...ViewStylePropTypes, ...TextStylePropTypes }),
        PropTypes.number,
      ])
    ),
    PropTypes.number,
  ]),

  name: PropTypes.string,
  resizingConstraint: PropTypes.shape({
    ...ResizingConstraintPropTypes,
  }),
  children: PropTypes.node,
};

/**
 * @example
 * <Text name='Foo' style={style}>
 *   Hello World!
 * </Text>
 */
class Text extends React.Component {
  static defaultProps = {
    name: 'Text',
  };

  render() {
    return (
      <text
        name={this.props.name}
        style={StyleSheet.flatten(this.props.style)}
        resizingConstraint={this.props.resizingConstraint}
      >
        {this.props.children}
      </text>
    );
  }
}

Text.propTypes = propTypes;

module.exports = Text;
