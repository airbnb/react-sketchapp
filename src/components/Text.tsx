import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import { StyleSheet } from '../stylesheet';
import { TextStylePropTypes } from './TextStylePropTypes';
import { ViewPropTypes } from './View';

export const TextPropTypes = {
  ...ViewPropTypes,
  style: or([PropTypes.shape(TextStylePropTypes), PropTypes.number]),
};

export type Props = PropTypes.InferProps<typeof TextPropTypes>;

/**
 * @example
 * <Text name='Foo' style={style}>
 *   Hello World!
 * </Text>
 */
export class Text extends React.Component<Props> {
  static propTypes = TextPropTypes;

  render() {
    return (
      <sketch_text
        name={this.props.name}
        style={StyleSheet.flatten(this.props.style)}
        resizingConstraint={this.props.resizingConstraint}
      >
        {this.props.children}
      </sketch_text>
    );
  }
}
