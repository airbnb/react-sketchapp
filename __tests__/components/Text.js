import React from 'react';
import renderer from 'react-test-renderer';
import Text from '../../src/components/Text';
import StyleSheet from '../../src/stylesheet';

describe('<Text />', () => {
  it('passes its children', () => {
    const tree = renderer.create(<Text>foo</Text>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('name', () => {
    it('passes its name', () => {
      const tree = renderer.create(<Text name="foo" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('defaults to Text', () => {
      const tree = renderer.create(<Text />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('style', () => {
    const styles = StyleSheet.create({
      view: {
        flex: 1,
      },
    });

    it('accepts a plain object', () => {
      const tree = renderer.create(<Text style={{ flex: 1 }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts a StyleSheet ordinal', () => {
      const tree = renderer.create(<Text style={styles.view} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts an array of plain objects and/or StyleSheet ordinals', () => {
      const tree = renderer
        .create(<Text style={[{ flexGrow: 1 }, styles.view]} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
