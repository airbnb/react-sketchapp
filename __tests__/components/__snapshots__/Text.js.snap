exports[`<Text /> name defaults to Text 1`] = `
<text
  name="Text"
  resizingConstraint={undefined}
  style={undefined} />
`;

exports[`<Text /> name passes its name 1`] = `
<text
  name="foo"
  resizingConstraint={undefined}
  style={undefined} />
`;

exports[`<Text /> passes its children 1`] = `
<text
  name="Text"
  resizingConstraint={undefined}
  style={undefined}>
  foo
</text>
`;

exports[`<Text /> style accepts a StyleSheet ordinal 1`] = `
<text
  name="Text"
  resizingConstraint={undefined}
  style={
    Object {
      "flex": 1,
    }
  } />
`;

exports[`<Text /> style accepts a plain object 1`] = `
<text
  name="Text"
  resizingConstraint={undefined}
  style={
    Object {
      "flex": 1,
    }
  } />
`;

exports[`<Text /> style accepts an array of plain objects and/or StyleSheet ordinals 1`] = `
<text
  name="Text"
  resizingConstraint={undefined}
  style={
    Object {
      "flex": 1,
      "flexGrow": 1,
    }
  } />
`;
